assets/aml-benchmark/components/src/prompt_crafter/package/prompt_factory.py

```diff
@@ -0,0 +1,340 @@
+            self.validate_jinja_template(self.augmented_few_shot_pattern,
+                                         "Few shot pattern is not a valid jinja pattern.")
+
+    def validate_jinja_template(self, template: str, error_message: str):
+        """Validate a jinja template."""
+        try:
+            _ = JINJA_ENV.from_string(template)
+        except Exception:
+            mssg = f"{template} is not a valid jinja pattern. Error: {error_message}"
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg))
+
+    @classmethod
+    def from_type(cls, prompt_type: PromptType):
+        """Create a prompt factory from a prompt type."""
+        if prompt_type == PromptType.completions.name:
+            return CompletionsPromptFactory
+        elif prompt_type == PromptType.chat.name:
+            return ChatPromptFactory
+        else:
+            raise ValueError(f"Unrecognized prompt type {prompt_type}. Should be \
+                             one of {PromptType.chat.name} or {PromptType.completions.name}")
+
+    @abstractmethod
+    def create_prompt(row: Dict) -> Prompt:
+        """Create prompts from a row of data."""
+        pass
+
+    @staticmethod
+    def _parse_label_map(label_map_str: str) -> Dict:
+        """Parse label map string into a dictionary."""
+        try:
+            label_map = json.loads(label_map_str)
+            label_map = {int(k): v for k, v in label_map.items()}
+        except Exception:
+            mssg = f"Invalid label map string:\n{label_map_str}"
+            logger.exception(mssg)
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg))
+        return label_map
+
+    @staticmethod
+    def _create_label_map_jinja_prefix(label_map_str: str) -> str:
+        label_map = PromptFactory._parse_label_map(label_map_str)
+        label_map_jinja_prefix = f"{{% set label_map = {label_map} %}}"
+        logger.info(f"Created label map jinja prefix:\n{label_map_jinja_prefix}")
+        try:
+            _ = JINJA_ENV.from_string(label_map_jinja_prefix)
+        except Exception:
+            mssg = "Label map is not a valid jinja template."
+            raise BenchmarkValidationException._with_error(
+                AzureMLError.create(BenchmarkValidationError, error_details=mssg))
+        return label_map_jinja_prefix
+
+    @staticmethod
+    def _inverse_label_map(label_map_str: str) -> Dict:
+        label_map = PromptFactory._parse_label_map(label_map_str)
+        return {value: key for key, value in label_map.items()}
+
+    def _sample_few_shots(self, row: Dict[str, str]):
+        few_shots = []
+        few_shot_ids = set()
+
+        retries = 0
+        while len(few_shots) < self.n_shots:
+
+            retries += 1
+            if retries > PATIENCE:
+                raise ValueError(
```

Comment:
I see a lot of `ValueError` and `RuntimeError` in multiple files. We should try raising exception that inherit from `UserException` class. Otherwise, these will be raised as `SystemError`.

Reply:
Removed. Thanks